akka-docs/rst/scala/code/docs/persistence/PersistenceSchemaEvolutionDocSpec.scala

```diff
@@ -0,0 +1,419 @@
+        case SeatReservedManifest =>
+          seatReserved(proto.SeatReserved.parseFrom(bytes)) // use generated protobuf serializer
+        case _ =>
+          throw new IllegalArgumentException("Unable to handle manifest: " + manifest)
+      }
+
+    override def toBinary(o: AnyRef): Array[Byte] = o match {
+      case s: SeatReserved =>
+        proto.SeatReserved.newBuilder
+          .setCode(s.letter)
+          .setSeatType(s.seatType)
+          .build().toByteArray
+    }
+
+    // -- fromBinary helpers --
+
+    private def seatReserved(p: proto.SeatReserved): SeatReserved =
+      SeatReserved(p.getLetter, p.getRow, seatType(p))
+
+    // handle missing field by assigning "Unknown" value
+    private def seatType(p: proto.SeatReserved): SeatType =
+      if (p.hasSeatType) SeatType.fromString(p.getSeatType) else SeatType.Unknown
+
+  }
+  //#protobuf-read-optional
+}
+
+class ProtoBufRename {
+  val protoIDL = """
+    //#protobuf-rename-proto
+    // protobuf message definition, BEFORE:
+    message SeatReserved {
+      required string code = 1;
+    }
+
+    // protobuf message definition, AFTER:
+    message SeatReserved {
+      required string seatNr = 1; // field renamed, id remains the same
+    }
+    //#protobuf-rename-proto
+  """
+}
+
+class RenamePlainJson {
+  //#rename-plain-json
+  class JsonRenamedFieldAdapter extends EventAdapter {
+    val marshaller = new ExampleJsonMarshaller
+
+    val V1 = "v1"
+    val V2 = "v2"
+
+    // this could be done independently for each event type
+    override def manifest(event: Any): String = "v2"
+
+    override def toJournal(event: Any): JsObject =
+    marshaller.toJson(event)
+
+    override def fromJournal(event: Any, manifest: String): EventSeq = event match {
+      case json: JsObject => EventSeq(marshaller.fromJson(manifest match {
+        case V1      => rename(json, "code" -> "seatNr")
+        case V2      => json // pass-through
+        case unknown => throw new IllegalArgumentException(s"Unknown manifest: $unknown")
+      }))
+      case _ =>
+        val c = event.getClass
+        throw new IllegalArgumentException("Can only work with JSON, was: %s".format(c))
+    }
+
+    def rename(json: JsObject, fromTo: (String, String)): JsObject = {
+      val value = json.fields(fromTo._1)
+      val withoutOld = json.fields - fromTo._1
+      JsObject(withoutOld + (fromTo._2 -> value))
+    }
+
+  }
+  //#rename-plain-json
+}
+
+//#simplest-custom-serializer-model
+final case class Person(name: String, surname: String)
+//#simplest-custom-serializer-model
+
+//#simplest-custom-serializer
+/**
+ * Simplest possible serializer, uses a string representation of the Person class.
+ *
+ * Usually a serializer like this would use a library like:
+ * protobuf, kryo, avro, cap'n proto, flatbuffers, SBE or some other dedicated serializer backend
+ * to perform the actual to/from bytes marshalling.
+ */
+class SimplestPossiblePersonSerializer extends Serializer {
```

Comment:
I'd prefer that this sample used SerializerWithStringManifest, because we always recommend that over the other Serializer for Akka Persistence. Users copy paste without reading.
This change can be done in a separate PR.